compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java

```diff
@@ -3467,30 +3467,44 @@ public BType resolveExprType(BType type) {
         }
     }
 
-    /**
-     * Check whether a order-key expression type is a basic type.
-     *
-     * @param type type of the order-key expression.
-     * @return boolean whether the type is basic type or not.
-     */
-    public boolean isBasicType(BType type) {
-        switch (type.tag) {
-            case TypeTags.INT:
+    private boolean isSimpleBasicType(int tag) {
+        switch (tag) {
             case TypeTags.BYTE:
             case TypeTags.FLOAT:
             case TypeTags.DECIMAL:
-            case TypeTags.STRING:
             case TypeTags.BOOLEAN:
             case TypeTags.NIL:
                 return true;
+            default:
+                return (TypeTags.isIntegerTypeTag(tag)) || (TypeTags.isStringTypeTag(tag));
+        }
+    }
+
+    /**
+     * Check whether a type is an ordered type.
+     *
+     * @param type type.
+     * @return boolean whether the type is an ordered type or not.
+     */
+    public boolean isOrderedType(BType type) {
+        switch (type.tag) {
             case TypeTags.UNION:
-                if (((BUnionType) type).getMemberTypes().contains(symTable.nilType)) {
-                    return true;
-                } else {
-                    return false;
+                Set<BType> memberTypes = ((BUnionType) type).getMemberTypes();
+                // can not sort (string?|int)/(string|int), can sort string?
+                return memberTypes.size() <= 2 && memberTypes.contains(symTable.nilType);
```

Comment:
What happens here if the union has a non-ordered type as a member type?

Reply:
Should be calling `isOrderedType()` recursively on the member types, in addition to the above checks right?